File: elevate/lib/redis/telemetry.ts
import type { CustomerEvent, TelemetrySnapshot, Anomaly } from '@/types'
import { getRedisClient, KEYS, TTL } from '@/lib/redis/client'

// ─── Telemetry Snapshot Layer ─────────────────────────────────────────────────

export async function recordEvent(
  merchantId: string,
  event: CustomerEvent
): Promise<void> {
  const redis = getRedisClient()

  // Track product velocity (views per product in sliding window)
  if (event.eventType === 'view') {
    await redis.zincrby(KEYS.productVelocity(merchantId), 1, event.productId)
    await redis.expire(KEYS.productVelocity(merchantId), 300) // 5 min window
  }

  // Store session event
  await redis.lpush(
    KEYS.sessionEvents(event.sessionId),
    JSON.stringify(event)
  )
  await redis.expire(KEYS.sessionEvents(event.sessionId), TTL.session)

  // Increment active session count
  await redis.sadd(`elevate:${merchantId}:active_sessions`, event.sessionId)
  await redis.expire(`elevate:${merchantId}:active_sessions`, 300)
}

export async function captureSnapshot(merchantId: string): Promise<TelemetrySnapshot> {
  const redis = getRedisClient()

  // Get active sessions
  const activeSessions = await redis.scard(`elevate:${merchantId}:active_sessions`)

  // Get product velocity scores
  const velocityData = await redis.zrevrange(
    KEYS.productVelocity(merchantId),
    0,
    -1,
    'WITHSCORES'
  )

  const productVelocity: Record<string, number> = {}
  for (let i = 0; i < velocityData.length; i += 2) {
    productVelocity[velocityData[i]] = parseFloat(velocityData[i + 1])
  }

  const hotProducts = velocityData
    .filter((_, i) => i % 2 === 0)
    .slice(0, 5)

  // Detect anomalies
  const anomalies = detectAnomalies(productVelocity, activeSessions)

  const snapshot: TelemetrySnapshot = {
    capturedAt: Date.now(),
    activeSessionCount: activeSessions,
    productVelocity,
    transactionRate: 0, // populated by purchase events
    abandonRate: 0,     // populated by session analysis
    hotProducts,
    anomalies,
  }

  // Cache snapshot
  await redis.set(
    KEYS.snapshot(merchantId),
    JSON.stringify(snapshot),
    'EX',
    TTL.snapshot
  )

  return snapshot
}

// ─── Anomaly detection ────────────────────────────────────────────────────────

function detectAnomalies(
  productVelocity: Record<string, number>,
  activeSessions: number
): Anomaly[] {
  const anomalies: Anomaly[] = []
  const velocities = Object.values(productVelocity)

  if (velocities.length === 0) return anomalies

  const avgVelocity = velocities.reduce((a, b) => a + b, 0) / velocities.length

  for (const [productId, velocity] of Object.entries(productVelocity)) {
    // Velocity spike: 3x above average
    if (velocity > avgVelocity * 3) {
      anomalies.push({
        type: 'velocity_spike',
        productId,
        severity: velocity > avgVelocity * 5 ? 'high' : 'medium',
        detectedAt: Date.now(),
        context: { velocity, avgVelocity, ratio: velocity / avgVelocity },
      })
    }

    // Dead product: zero velocity while others are active
    if (velocity === 0 && activeSessions > 5) {
      anomalies.push({
        type: 'dead_product',
        productId,
        severity: 'low',
        detectedAt: Date.now(),
        context: { activeSessions },
      })
    }
  }

  return anomalies
}

File: analytics-brain/app/services/telemetry.py
import time
from app.core.redis import get_redis, Keys, TTL
from app.models.schemas import CustomerEvent, TelemetrySnapshot, Anomaly, AnomalyType


async def record_event(merchant_id: str, event: CustomerEvent) -> None:
    redis = await get_redis()

    if event.event_type.value == "view":
        await redis.zincrby(Keys.product_velocity(merchant_id), 1, event.product_id)
        await redis.expire(Keys.product_velocity(merchant_id), TTL.SNAPSHOT)

    await redis.sadd(Keys.active_sessions(merchant_id), event.session_id)
    await redis.expire(Keys.active_sessions(merchant_id), TTL.SNAPSHOT)

    await redis.lpush(
        Keys.session_events(event.session_id),
        event.model_dump_json(),
    )
    await redis.expire(Keys.session_events(event.session_id), TTL.SESSION)


async def capture_snapshot(merchant_id: str) -> TelemetrySnapshot:
    redis = await get_redis()

    active_sessions = await redis.scard(Keys.active_sessions(merchant_id))

    velocity_data = await redis.zrevrange(
        Keys.product_velocity(merchant_id), 0, -1, withscores=True
    )

    product_velocity: dict[str, float] = {
        product_id: score for product_id, score in velocity_data
    }
    hot_products = [pid for pid, _ in velocity_data[:5]]
    anomalies = _detect_anomalies(product_velocity, active_sessions)

    snapshot = TelemetrySnapshot(
        captured_at=int(time.time() * 1000),
        active_session_count=active_sessions,
        product_velocity=product_velocity,
        hot_products=hot_products,
        anomalies=anomalies,
    )

    await redis.set(
        Keys.snapshot(merchant_id),
        snapshot.model_dump_json(),
        ex=TTL.SNAPSHOT,
    )

    return snapshot


def _detect_anomalies(
    product_velocity: dict[str, float],
    active_sessions: int,
) -> list[Anomaly]:
    anomalies: list[Anomaly] = []
    velocities = list(product_velocity.values())
    if not velocities:
        return anomalies

    avg = sum(velocities) / len(velocities)
    now = int(time.time() * 1000)

    for product_id, velocity in product_velocity.items():
        if velocity > avg * 3:
            anomalies.append(Anomaly(
                type=AnomalyType.VELOCITY_SPIKE,
                product_id=product_id,
                severity="high" if velocity > avg * 5 else "medium",
                detected_at=now,
                context={"velocity": velocity, "avg": avg, "ratio": round(velocity / avg, 2)},
            ))
        if velocity == 0 and active_sessions > 5:
            anomalies.append(Anomaly(
                type=AnomalyType.DEAD_PRODUCT,
                product_id=product_id,
                severity="low",
                detected_at=now,
                context={"active_sessions": active_sessions},
            ))

    return anomalies
